Add the Go version used to build the release to the release announcement

xmudrii opened the pull request:
#### What type of PR is this?

/kind bug regression

#### What this PR does / why we need it:

This PR adds the Go version used to build the release to the release announcement.

This is fixing a regression, as originally we had the Go version in the release announcement, but it appears that we lost it in some of the recent refactors.

The implemented `getGoVersion` function runs a kube-cross container and then runs `go version` inside it. We do it because it's not guaranteed that the Go version inside the k8s-cloud-builder container (which runs krel) is the same Go version used in the kube-cross container which builds the release.

#### Which issue(s) this PR fixes:

Addresses two action items from https://github.com/kubernetes/sig-release/issues/1365:

- Announcement is missing the Go version used to build the release
- The Kubernetes version and Go versions were bold in the old template

#### Does this PR introduce a user-facing change?

```release-note
Ensure the Go version used to build the release is present in the announcement
```

/priority important-soon
/assign @puerco @saschagrunert 

/hold
for discussion

saschagrunert commented on pkg/anago/release.go:
```diff
@@ -525,6 +535,20 @@ func (d *DefaultRelease) CreateAnnouncement() error {
+func (d *defaultReleaseImpl) GoVersion() (string, error) {
+	const semVerRegex = `^?([0-9]+)(\.[0-9]+)?(\.[0-9]+)`
+
+	cmdStatus, err := command.New("go", "version").RunSilentSuccessOutput()
```
I think we can use `runtime.Version()` instead of the command execution.

xmudrii replied:
That's a good point. I'll update the PR.

saschagrunert commented on pkg/announce/options.go:
```diff
@@ -69,3 +72,8 @@ func (o *Options) WithChangelogFile(changelogFile string) *Options {
 	return o
 }
+
+func (o *Options) WithGoVersion(goVersion string) *Options {
+	o.goVersion = goVersion
+	return o
+}
```
Do we want to add the version always? I don't see a strong reason for making it optional.

xmudrii replied:
I did it because of the consistency. We pass all other variables via Options struct, so I thought it would eventually make sense to do the same for the Go version.

However, the Go version is not really dynamic, unlike other properties, so maybe I should just do it from `pkg/announce/announce.go` instead of passing it as an option.

xmudrii commented:
@saschagrunert @puerco I've refactored the PR a bit, please take a look.

:hourglass_flowing_sand: Mock Release: https://console.cloud.google.com/cloud-build/builds/5ced5c48-616f-41fb-b29e-31ae3e4d48c1?project=kubernetes-release-test

xmudrii commented on pkg/announce/announce.go:
```diff
@@ -143,3 +154,33 @@ func create(workDir, subject, message string) error {
+
+	releaseBranch := fmt.Sprintf("release-%d.%d", semver.Major, semver.Minor)
+
+	kubecrossVer, err := release.GetKubecrossVersion(releaseBranch)
```
I'm considering to include the master branch here as well, such as:

```suggestion
	kubecrossVer, err := release.GetKubecrossVersion(releaseBranch, "master")
```

(I think there might be something like `git.DefaultBranch` or something like that, instead of hard-coded `master`)

I think that it might happen that we cut a release that still doesn't have the kube-cross variant (that should be all releases before and including rc.0). In that case, this function would fail, but if we add the master branch, it would have a fall-back.

puerco replied:
Yep, we use `git.DefaultBranch` everywhere (I see you used it, nice :)

xmudrii commented:
Addressed the comment I left.

:hourglass_flowing_sand: Mock Release: https://console.cloud.google.com/cloud-build/builds/3d5d9dcb-e598-4782-8760-3955a2649ca7?project=kubernetes-release-test

puerco commented:
Looks fine!

```html
Kubernetes <b>v1.20.0-rc.1</b> has been built and pushed using Golang version <b>1.15.5</b>.
```

puerco commented on pkg/announce/announce.go:
```diff
@@ -143,3 +155,36 @@ func create(workDir, subject, message string) error {
+
+	kubecrossImg := fmt.Sprintf("k8s.gcr.io/build-image/kube-cross:%s", kubecrossVer)
+
+	res, err := command.New(
+		"docker", "run", "--rm", kubecrossImg, "go", "version",
+	).RunSilentSuccessOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get go version")
+	}
```
Nice, now there is no chance of misreporting :+1:

puerco commented on pkg/announce/announce.go:
```diff
@@ -143,3 +155,36 @@ func create(workDir, subject, message string) error {
+	}
+
+	versionRegex := regexp.MustCompile(`^?([0-9]+)(\.[0-9]+)?(\.[0-9]+)`)
+	return versionRegex.FindString(strings.TrimSpace(res.OutputTrimNL())), nil
```
Just a nitty nit, I would log the go version we found, it is this code's info we want to record in the logs in case we need to go back to them

puerco replied:
can be left for a follow up though :)

xmudrii replied:
@puerco Fixed, PTAL. :slightly_smiling_face:

puerco commented:
It looks fine to me, thanks for handling this one marko :) 
/lgtm

xmudrii commented on pkg/announce/announce.go:
```diff
@@ -104,13 +111,20 @@ func CreateForRelease(opts *Options) error {
 		changelog = opts.changelogHTML
 	}
 
+	logrus.Infof("Trying to get the Go version used to build %s...", opts.tag)
+	goVersion, err := getGoVersion(opts.tag)
+	if err != nil {
+		return err
+	}
+	logrus.Infof("Found the following Go version: %s", goVersion)
```
@puerco Maybe I should have put this in the `getGoVersion` function?

puerco replied:
Oh sorry I did not see it in the log, that's why I left the comment:
```
Step #2: INFO[2020-12-04T15:49:03Z] Creating v1.20.0-rc.1 announcement in /workspace/src 
Step #2: INFO[2020-12-04T15:49:03Z] Trying to get the kube-cross version for release-1.20... 
Step #2: INFO[2020-12-04T15:49:03Z] Found the following kube-cross version: v1.15.5-1 
Step #2: INFO[2020-12-04T15:49:04Z] Release announcement created           
```

xmudrii replied:
@puerco I've actually added it after your comment. :slightly_smiling_face:

puerco commented:
/lgtm

saschagrunert reviewed (approved):
/lgtm

xmudrii commented:
@saschagrunert @puerco Thank you for the review! Is it okay if remove the hold from the PR?

puerco commented:
Do it !  :rocket:

xmudrii commented:
/hold cancel
